### PR TITLE
feat(layer-cache): adds layerdb abstraction

### PR DESCRIPTION
**Pull request by adamhjk**
* Adds a layerdb abstraction
* Adds a persister thread
* Switches to a table-per-object rather than a single large table
* Extends larger events

Written by Zack, Scott, Fletcher and Adam.

### CODE REVIEW

**Comment by zacharyhamm** [comment]
bors merge

File: lib/si-layer-cache/tests/integration_test/disk_cache.rs [hunk 1]
```diff
@@ -2,15 +2,15 @@ use si_layer_cache::disk_cache::DiskCache;
 
 #[test]
 fn new() {
-    let tempdir = tempfile::tempdir().expect("cannot create tempdir");
+    let tempdir = tempfile::TempDir::new_in("/tmp").expect("cannot create tempdir");
```
Comment (nickgerace) [inline lib/si-layer-cache/tests/integration_test/disk_cache.rs:5]: Did `tempdir()` not behave as expected?

File: lib/si-events-rs/src/cas.rs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                        .expect("serde_json said it was an i64 but refused to give me one"),
+                )
+            } else if n.is_f64() {
+                CasValueNumber::F64(
+                    n.as_f64()
+                        .expect("serde_json said it was an f64 but refused to give me one"),
+                )
+            } else {
+                panic!("the arbitrary_precision feature of serde_json is not supported");
+            }),
```
Comment (nickgerace) [inline lib/si-events-rs/src/cas.rs:82]: Would `TryFrom` and returning the error be valuable here?

File: lib/si-events-rs/src/cas.rs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            CasValue::Null => serde_json::Value::Null,
+            CasValue::Bool(b) => serde_json::Value::Bool(b),
+            CasValue::Array(mut a) => {
+                serde_json::Value::Array(a.drain(..).map(Into::into).collect())
+            }
+            CasValue::Number(n) => serde_json::Value::Number(match n {
+                CasValueNumber::U64(n) => n.into(),
+                CasValueNumber::I64(n) => n.into(),
+                CasValueNumber::F64(n) => serde_json::value::Number::from_f64(n)
+                    .expect("cannot deserialize an infinite or NAN f64 value"),
```
Comment (nickgerace) [inline lib/si-events-rs/src/cas.rs:107]: Same question here as above. I assume it is impossible to hit though.

File: lib/si-layer-cache/src/persister.rs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+impl PersisterStatusWriter {
+    pub fn new(tx: oneshot::Sender<PersistStatus>) -> Self {
+        Self { tx }
+    }
+
+    pub fn send(self, msg: PersistStatus) {
+        // If the other end isn't listening, we really don't care!
+        let _ = self.tx.send(msg);
+    }
```
Comment (nickgerace) [inline lib/si-layer-cache/src/persister.rs:41]: Gotcha

File: lib/si-layer-cache/tests/integration_test/db/cas.rs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        .await
+        .expect("error getting data from pg")
+        .expect("no cas object in pg");
+    let in_pg: CasValue =
+        postcard::from_bytes(&in_pg_postcard[..]).expect("cannot deserialize data");
+    assert_eq!(cas_value.as_ref(), &in_pg);
+}
+
+#[tokio::test]
+async fn cold_read_from_db() {
```
Comment (nickgerace) [inline lib/si-layer-cache/tests/integration_test/db/cas.rs:68]: Makes sense to me! Good test.

**Pull request merged**